controller_manager/src/controller_manager.cpp

```diff
@@ -137,6 +128,141 @@ void ControllerManager::getControllerNames(std::vector<std::string> &names)
   }
 }
 
+void ControllerManager::manageSwitch(const ros::Time &time)
+{
+  // switch hardware interfaces (if any)
+  if (!switch_started_)
+  {
+    robot_hw_->doSwitch(switch_start_list_, switch_stop_list_);
+    switch_started_ = true;
+  }
+
+  stopControllers(time);
+
+  // start controllers once the switch is fully complete
+  if (!start_asap_)
+  {
+    startControllers(time);
+  }
+  // start controllers as soon as their required joints are done switching
+  else
+  {
+    startControllersAsap(time);
+  }
+}
+
+void ControllerManager::stopControllers(const ros::Time &time)
+{
+  // stop controllers
+  for (auto &request : stop_request_)
+  {
+    if (request->isRunning())
+    {
+      if (!request->stopRequest(time))
+      {
+        ROS_FATAL("Failed to stop controller in realtime loop. This should never happen.");
+      }
+    }
+  }
+}
+
+void ControllerManager::startControllers(const ros::Time &time)
+{
+  // start controllers
+  if (robot_hw_->switchResult() == hardware_interface::RobotHW::DONE)
+  {
+    for (auto &request : start_request_)
+    {
+      if (!request->startRequest(time))
+      {
+        ROS_FATAL("Failed to start controller in realtime loop. This should never happen.");
+      }
+    }
+
+    please_switch_ = false;
+    switch_started_ = false;
+  }
+  // abort controllers in case of error or timeout (if set)
+  else if ((robot_hw_->switchResult() == hardware_interface::RobotHW::ERROR) ||
+           (timeout_ > 0.0 && (time - init_switch_).toSec() > timeout_))
+  {
+    for (auto &request : start_request_)
+    {
+      if (!request->abortRequest(time))
+      {
+        ROS_FATAL("Failed to abort controller in realtime loop. This should never happen.");
```

Comment:
"This should never happen" may not carry a whole lot of value to the one who reads these messages :) Can we remove them or move into a comment instead please?

Reply:
Frankly, I don't like it either. I was just following the pattern of the original implementation.

```C++
    // stop controllers
    for (unsigned int i=0; i<stop_request_.size(); i++)
      if (!stop_request_[i]->stopRequest(time))
        ROS_FATAL("Failed to stop controller in realtime loop. This should never happen.");

    // start controllers
    for (unsigned int i=0; i<start_request_.size(); i++)
      if (!start_request_[i]->startRequest(time))
        ROS_FATAL("Failed to start controller in realtime loop. This should never happen.");
```

By design, these calls can fail only if you are calling an uninitialized controller. So perhaps we could change the error message to:

```C++
  ROS_FATAL("Failed to stop controller in realtime loop. Controller is not initialized");
```
What do you think?

Reply:
It would more correct for the message to be logged by the `ControllerBase` also...

Reply:
I completely agree with both points!

Reply:
Done!